File: CharacterStream.cpp
class StreamState {
public:
  int index;
  bool ended;
  Location location;
  StreamState(int index, bool ended, Location location) :
    index(index), ended(ended), location(location) {}
};
class StreamException : public SolException {
public:
  StreamException(string s) :
    SolException("Stream", s) {}
};
/*abstract*/class CharacterStream {
public:
  int index;
  bool ended;
  bool closed;
  Location location;
  CharacterStream(string filename) :
    location(filename), index(0), ended(false), closed(false){}
  CharacterStream(Location location) :
    location(location), index(0), ended(false), closed(false) {}
  bool isEOF() { return ended; }
  bool EOFState() { ended = true; }
  void close() { closed = true; closeSpecialized(); }
  virtual char getCharacterSpecialized() = 0;
  virtual void nextCharacterSpecialized() = 0;
  virtual void restoreStateSpecialized(StreamState *state) = 0;
  virtual void closeSpecialized() {};
  char getCharacter() {
    if(isEOF())
      return EOF;
    else
      return getCharacterSpecialized();
  }
  void nextCharacter() {
    if(isEOF())
      return;
    if(getCharacter() == '\n')
      location.advanceNextLine();
    else
      location.advanceNextCharacter();
    index++;
    nextCharacterSpecialized();
  }
  StreamState *saveState() {
    if(closed)
      throw StreamException("Attempting to save a closed stream state.");
    return new StreamState(index, ended, location);
  }
  void restoreState(StreamState *state) {
    if(closed)
      throw StreamException("Attempting to restore a closed stream state.");
    index = state->index;
    ended = state->ended;
    location = state->location;
    restoreStateSpecialized(state);
  }
  Location currentLocation() { return location; }
};

class StringStream : public CharacterStream {
public:
  string line;
  StringStream(string filename, string line) :
    CharacterStream(filename), line(line) {}
  char getCharacterSpecialized() {
    return line[index];
  }
  void nextCharacterSpecialized() {
    if(index >= line.length())
      EOFState();
  }
  void restoreStateSpecialized(StreamState *state) {}
  void close() {}
};

class FileStream : public StringStream {
public:
  string filename;
  string content;
  FileStream(string filename) :
    StringStream(filename, readWholeFile(filename)) {}

  static string readWholeFile(string filename) {
    ifstream file(filename);
    if(!file.good())
      throw StreamException("Unable to open \"" + filename + "\".");
    stringstream buffer;
    buffer << file.rdbuf();
    file.close();
    return buffer.str();
  }
};

class MultilineReplStream : public CharacterStream {
public:
  string prompt;
  string multilinePrompt;
  string content;
  string lastLine;
  MultilineReplStream(string prompt) :
    CharacterStream("<input>"), prompt(prompt) {
    nextCharacterSpecialized();
  }
  char getCharacterSpecialized() {
    return content[index];
  }
  void nextCharacterSpecialized() {
    if(index >= content.length()) {
      cout<<prompt;
      getline(cin, lastLine);
      if(lastLine == "") 
	EOFState();
      else
	content += lastLine + '\n';
    }
  }
  void restoreStateSpecialized(StreamState *state) {
    lastLine = "";
  }
  void close() {}
};

File: src/CharacterStream.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public abstract class CharacterStream {
    int index;
    boolean eof;
    Location location;

    class State {
        public int index;
        public boolean eof;
        public Location location;
        State(int index, boolean eof, Location location) {
            this.index = index;
            this.eof = eof;
            this.location = new Location(location);
        }
    }

    CharacterStream(String filename) {
        this.location = new Location(filename);
    }
    public boolean hasCharacters() { return !eof; }
    public Location getLocation() { return location; }
    public void EOFState() { eof = true; }
    public abstract void close();
    public abstract char getCharacterSpecialized();
    public abstract void nextCharacterSpecialized();
    public abstract void restoreStateSpecialized(State state);
    public State saveState() {
        return new State(index, eof, location);
    }
    public void restoreState(State state) {
        index    = state.index;
        eof      = state.eof;
        location = state.location;

        restoreStateSpecialized(state);
    }
    public char getCharacter() {
        return !hasCharacters() ? 0 : getCharacterSpecialized();
    }
    public void nextCharacter() {
        if(!hasCharacters())
            return;
        if(getCharacter() == '\n')
            location.advanceNextLine();
        else
            location.advanceNextCharacter();
        index++;
        nextCharacterSpecialized();
    }
}

class StringStream extends CharacterStream {
    String line;
    StringStream(String line) {
        super("<input>");
        this.line = line;
        if(this.line.length() == 0)
            EOFState();
    }
    StringStream(String filename, String line) {
        super(filename);
        this.line = line;
        if(this.line.length() == 0)
            EOFState();
    }
    public char getCharacterSpecialized() { return line.charAt(index); }
    public void nextCharacterSpecialized() {
        if(index >= line.length())
            EOFState();
    }
    public void restoreStateSpecialized(State state) {}
    public void close() {}
}

class FileStream extends StringStream {
    String line;
    String content;
    FileStream(String filename) throws IOException {
        super(filename, readWholeFile(filename));
    }
    static String readWholeFile(String filename) throws IOException {
        return new String(Files.readAllBytes(Paths.get(filename)));
    }
}

class MultilineReplStream extends CharacterStream {
    String prompt;
    String multilinePrompt;
    String content;
    String lastLine;
    MultilineReplStream(String prompt) {
        super("<input>");
        this.content = "";
        this.lastLine = "";
        this.prompt = prompt;
        nextCharacterSpecialized();
    }
    public char getCharacterSpecialized() { return content.charAt(index); }
    public void nextCharacterSpecialized() {
        if(index >= content.length()) {
            try {
                System.out.print(prompt);
                BufferedReader bufferedIn = new BufferedReader(new InputStreamReader(System.in));
                lastLine = bufferedIn.readLine();
                if(lastLine == null)
                    EOFState();
                else
                    content += lastLine + '\n';
            } catch(IOException e) {
                EOFState();
            }
        }
    }
    public void restoreStateSpecialized(State state) { lastLine = ""; }
    public void close() {}
}
